packages/runtime/test-runtime-utils/src/insecureUrlResolver.ts

```diff
@@ -90,8 +90,9 @@ export class InsecureUrlResolver implements IUrlResolver {
         const encodedTenantId = encodeURIComponent(this.tenantId);
         const encodedDocId = encodeURIComponent(documentId);
         const host = new URL(this.ordererUrl).host;
-        const relativePath = !documentRelativePath || documentRelativePath.startsWith("/")
-            ? documentRelativePath : `/${documentRelativePath}`;
+        const tmpRelativePath = !documentRelativePath || documentRelativePath.startsWith("/")
+        ? documentRelativePath : `/${documentRelativePath}`;
+        const relativePath = tmpRelativePath === undefined ? "" : tmpRelativePath;
```

Comment:
This does not look like correct change. The typing clearly states that documentRelativePath can not be undefined. So there is no way for tmpRelativePath to be undefined.
If it is, that points to a bug in the caller, and the fix should be in the caller.

If it's hard to debug or we want system to be more defensive, then adding assert that tmpRelativePath !== undefined might be warranted, but as a general rule this is not advisable (as following same president, all the code would need to get these kinds of defensive checks

Reply:
private resolveHelper(documentId: string, **documentRelativePath: string**, queryParams: string = "") {
        ...
        const relativePath = !documentRelativePath || documentRelativePath.startsWith("/")
            ? documentRelativePath : `/${documentRelativePath}`;

It's type is string, not string | undefined, so compiler correctly tells me the result of this expression is string:

![image](https://user-images.githubusercontent.com/46826984/116320516-3658df80-a76d-11eb-8663-1031855393c9.png)

Reply:
<img width="1792" alt="Screen Shot 2021-04-27 at 3 56 48 PM" src="https://user-images.githubusercontent.com/15995082/116322716-7621c600-a771-11eb-88f6-745e2783d1e1.png">
<img width="1792" alt="Screen Shot 2021-04-27 at 3 56 39 PM" src="https://user-images.githubusercontent.com/15995082/116322772-93569480-a771-11eb-89bb-be4cc87f2471.png">
Please see the attached above, I use debug mode to run counterEndToEndTest and it shows that **documentRelativePath** and  **relativePath** is undefined.
So I think typing is not enough to guarantee that  documentRelativePath and relativePath won't be undefined

Reply:
Sure, what if it was a number? or an object? In all cases (including undefined) caller is responsible to adhere to contract which is defined by signature of the function. We should not assume on every line in repo that uses string that it can be string | undefined.
Please debug it and find the offender, i.e. code that violates contracts that results in undefined being passed in - that's the right place to fix, not this function. Should be easy once you have it in debugger.